File: javalin-openapi-processor/src/main/java/net/dzikoysk/openapi/processor/utils/JsonUtils.java
package net.dzikoysk.openapi.processor.utils;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;

import java.util.Arrays;
import java.util.function.Function;
import java.util.function.Supplier;

public final class JsonUtils {

    private JsonUtils() {}

    public static <T> JsonArray toArray(T[] array, Function<T, String> mapper) {
        JsonArray jsonArray = new JsonArray(array.length);

        Arrays.stream(array)
                .map(mapper)
                .forEach(jsonArray::add);

        return jsonArray;
    }

    public static JsonObject computeIfAbsent(JsonObject root, String key, Supplier<JsonObject> value) {
        if (!root.has(key)) {
            root.add(key, value.get());
        }

        return root.getAsJsonObject(key);
    }

}

File: javalin-openapi-processor/src/main/kotlin/net/dzikoysk/openapi/processor/utils/JsonUtils.kt
package net.dzikoysk.openapi.processor.utils

import com.google.gson.JsonArray
import com.google.gson.JsonObject
import java.util.function.Supplier

object JsonUtils {

    fun <T> toArray(list: List<T>, mapper: (T) -> String = { it.toString() }): JsonArray {
        val jsonArray = JsonArray(list.size)
        list.map(mapper).forEach { jsonArray.add(it) }
        return jsonArray
    }

    fun computeIfAbsent(root: JsonObject, key: String, value: Supplier<JsonObject>): JsonObject {
        if (!root.has(key)) {
            root.add(key, value.get())
        }

        return root.getAsJsonObject(key)
    }

}